base/src/Data/Macaw/Dwarf.hs

```diff
@@ -128,89 +138,96 @@ instance Monad m => WarnMonad String (WarnT m) where
 -- Parser
 
 -- | The context needed to read dwarf entries.
-data ParserState = ParserState { expectedPointerWidth :: !Word64
-                                 -- ^ Number of bytes a pointer is expected to have.
-                               , readerInfo :: !Dwarf.Reader
-                               }
+newtype ParserState = ParserState { readerInfo :: Dwarf.Reader
+                                  }
 
-newtype Parser r = Parser { unParser :: ReaderT ParserState (WarnT Identity) r }
+newtype Parser r = Parser { unParser :: ReaderT ParserState (WarnT String Identity) r }
   deriving ( Functor
            , Applicative
            , Monad
+           , MonadError String
            , WarnMonad String
-           , MF.MonadFail
            )
 
 
-runParser :: Word64 -> Dwarf.Reader -> Parser r -> (Either String r, [String])
-runParser w dr p = runIdentity (runWarnT (runReaderT (unParser p) s))
-  where s = ParserState { expectedPointerWidth = w
-                        , readerInfo = dr
+runParser :: Dwarf.Reader -> Parser r -> (Either String r, [String])
+runParser dr p = runIdentity (runWarnT (runReaderT (unParser p) s))
+  where s = ParserState { readerInfo = dr
                         }
 
 ------------------------------------------------------------------------
 -- Parser functions
 
+-- | Error from parsing an attribute.
+data AttrError
+   = IncorrectTypeFor !String
```

Comment:
Just a general comment - making a `String` strict only forces the cons cell of the first character to whnf

Reply:
Agreed.  I should stop using `String`.